Step5 : 리뷰 요청드립니다!

Mo-bile opened the pull request:
안녕하세요? 모재영입니다.

이번 미션을 하면서... 느낀점은 지난 과제까지 복잡해도 너무 복잡하게 해서 스스로를 난이도를 높여서 했다고 느껴졌었습니다.
보여주신 샘플코드를 읽고나서 이 흐름대로 접근해보고 구현하니 쉽게 구현이 되었습니다.
하면서도 일부 타당성검증이라던지 이런 부분은 생략했지만,,, 이렇게 쉽게 된다고? 하면서 스스로 생각이 들었습니다.

어떻게 처음에 클래스 설계를 하고 접근하느냐에 따라 방향성과 난이도 차이가 크게 벌어짐을 깨닫게 되었습니다....!

그러면서 고민한것은 `List<Car>` 를 일급컬렉션으로 하는 `JoinCars` 를 만들까 고민을 했습니다.
그러면은 이전 step 과제처럼 객체간의 복잡성은 올라가는것은 아닌지 하고 주춤하게되기도 합니다.
그래서 일급컬렉션을 만들었다가 지웠다가 반복하다가 결국에 `carsRace()`라는 메서드 분리정도만 하기로 했습니다!

step5 추가연습으로 깨달음을 얻게 해주셔서 감사드립니다...!

Commit:
add dependency junit-jupiter-params

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Commit:
refactor : 재연습을 위해 제거

Commit:
init : 초기구조 정의

Commit:
feat
1. Car 모델의 선언과 이동 구현 완료
2. CarFactory 로 출전 자동차 생성
3. RacingGame 게임에서 게임진행과 결과 리턴
4. RacingResult 게임 결과에 대해서 리턴하는 모델 구현 완료

Commit:
feat : 자동차 이동 시 랜덤성 추가

Commit:
test : 잘못된 조건 수정

Commit:
feat
1. 게임 기본 입출력 개발 완료 (우승자 선정제외)

Commit:
feat : 우승자 구하기 로직 개발 완료

Commit:
refactor
RacingResult 에 우승자구하는 로직에서 일부기능 Car로 메소드 이용으로 car에 행동을 줌

Commit:
refactor : List<Car>에 대한 일급컬렉션 생성대신에 carsRace() 메서드 분리를 해서 객체간 복잡도를 낮춤

javajigi commented on src/main/java/racinggame/domain/business/RacingGame.java:
```diff
@@ -0,0 +1,41 @@
+public class RacingGame {
+    
+    public static final int DEFAULT_COUNT = 1;
+    private int raceCount;
```
규칙 3: 모든 원시값과 문자열을 포장한다. 원칙 적용해 본다.

javajigi commented on src/main/java/racinggame/domain/business/RacingGame.java:
```diff
@@ -0,0 +1,41 @@
+    
+    public static final int DEFAULT_COUNT = 1;
+    private int raceCount;
+    private final List<Car> cars;
```
고민했다고 하니 "규칙 8: 일급 콜렉션을 쓴다." 원칙 적용해 구현해보고 리팩터링해본다.

javajigi commented on src/main/java/racinggame/domain/business/RacingGame.java:
```diff
@@ -0,0 +1,41 @@
+        carsRace();
+        this.raceCount --;
+        return new RacingResult(this.cars);
+    }
```
```suggestion
    public void race() {
        carsRace();
        this.raceCount --;
    }
    
    public RacingResult raceResult() {
        return new RacingResult(this.cars);
    }
```
재영님이 구현한 코드와 아래와 같이 분리해 구현했을 때 각각의 장단점은 무엇인가?
어떻게 구현하는 것이 좋을까 고민해 보는 계기가 되면 좋겠음.

javajigi commented on src/main/java/racinggame/domain/model/Car.java:
```diff
@@ -0,0 +1,43 @@
+    
+    public static final int INIT_LOCATION = 0;
+    private final String name;
+    private int location;
```
지금의 전체적인 객체 설계 구조하에서 "규칙 3: 모든 원시값과 문자열을 포장한다." 원칙을 적용해 가면서 객체를 분리하는 연습해보면 좋겠다.

javajigi commented on src/main/java/racinggame/domain/model/RacingResult.java:
```diff
@@ -0,0 +1,32 @@
+
+public record RacingResult(List<Car> cars) {
+    
+    public List<String> findWinners() {
```
List<Car>에 대해 일급 콜렉션을 적용하면서 우승자 구하는 로직이 어느 객체가 책임을 지는 것이 좋을지 고민해 보면 좋겠다.

javajigi reviewed (changes requested):
와. 재영님
제가 예시로 제공한 샘플 예제 코드 하나만으로 이렇게 깔끔하게 구현하다니 넘 칭찬하고 싶네요. 💯 
기존의 복잡한 구조에 비해 너무 단순하고 깔끔해 진 것 같아요.
이렇게 단순하게 구현한 코드를 바탕으로 객체를 하나씩 분리해 나가는 리팩터링 연습해 볼 것을 추천합니다.
이 상황이 우리가 현장에서 부딪치는 상황이거든요.
오늘 강의에서도 이 내용을 집중적으로 강의할텐데요.
미리 선 경험하고 수업 참여한다면 효과가 배가 될 것 같아요.
특히 List<Car>에 대해 일급 콜렉션 추가한 후 우승자 구하는 로직이 어디에 위치하는 것이 좋을지만이라도 도전해 보면 더 의미있을 것 같아요.

Commit:
refactor :
cars 라는 일급컬렉션을 만듬
1. CarFactory 객체제거 및 팩터리 메서드 Cars로 이관
2. RacingResult 에 있던 우승자 찾는 메서드 Cars로 이관

Commit:
refactor : Car 의 name 필드 객체 래핑

Commit:
refactor : Car의 Location 객체 분리
Car에 있던 위치 관련 메서드 내부 로직 Location으로 이관

Commit:
refactor :
RacingGame 에 있던 raceCount 필드를 객체로 분리

Commit:
fix : 종료조건 관련 부등호 수정

Commit:
fix : 잘못된 조건문에 not 추가

Commit:
refactor :
RacingGame 에 race() 라는 행동만하고 그 결과는 주지 않는 변경으로 이 메서드에 대한 예상을 가능하게 만듬 (CQS 패턴)
물론 기존처럼 행동 + 데이터 반환도 가능하지만 비즈니스 로직인 'RacingGame' 은 컨트롤러 레이어와 직접 붙는 부분이기 때문에 CQS패턴 채택이 클라이언트 코드 입장에서는 더 명확해서 채택했음

Commit:
doc : 체크리스트 완료

Commit:
refactor : 필드를 모두 객체로 분리 하여 불변 필드가 되어 record로 변경

Mo-bile commented:
안녕하세요 
리뷰 요청드립니다 👍 

직전 라이브 강의와 주신 피드백 바탕으로 객체 분리와 리팩터링을 시도했습니다
늘 감사드립니다.

아울러 아래 두개의 장단점에 대해서 고민할 것을 말씀주셨는데요
- before
```java
    public RacingResult race() {
        carsRace();
        this.raceCount --;
        return new RacingResult(this.cars);
    }
```

- after
```java
    public void race() {
        carsRace();
        this.raceCount --;
    }
    
    public RacingResult raceResult() {
        return new RacingResult(this.cars);
    }
```
저는 고민결과 후자를 택했습니다.

그 이유는 `RacingGame` 객체는 컨트롤러 레이어와 직접 붙어서 교류하는 객체이기 때문에
클라이언트 코드는 `race()` 이후에는 결과가 온다는 기대를 하기 어려울 수도 있기 때문이라 생각합니다..

`RacingGame` 에 `race()` 라는 '행동만하고 그 결과는 주지 않는 변경으로 이 메서드에 대한 예상을 가능하게 만들었습니다 (CQS 패턴) 즉 CQS패턴이 컨트롤러인 클라이언트 입장에서는 더 명확해서 채택

만약 내부 객체간 행동을 주고 받는다면 `행동 + 데이터 반환` 이 가능한 상황도 고민해볼 수있을 것 같습니다.

좋은 고민지점 주셔서 감사드립니다 😊

javajigi commented on src/main/java/racinggame/controller/RacingGameApplication.java:
```diff
@@ -0,0 +1,25 @@
+        while(racingGame.isEnd()) {
+            racingGame.race();
+            result = racingGame.raceResult();
+            ResultView.printAllResult(result);
```
```suggestion
            ResultView.printAllResult(racingGame.raceResult());
```

javajigi commented on src/main/java/racinggame/controller/RacingGameApplication.java:
```diff
@@ -0,0 +1,25 @@
+            result = racingGame.raceResult();
+            ResultView.printAllResult(result);
+        }
+        ResultView.winnersPrint(result);
```
```suggestion
        ResultView.winnersPrint(racingGame.findWinners());
```
위와 같이 우승자 목록을 직접 접근하는 것도 한 가지 방법일 수 있겠다.

javajigi commented on src/main/java/racinggame/domain/business/RacingGame.java:
```diff
@@ -0,0 +1,32 @@
+import racinggame.domain.model.RaceCount;
+import racinggame.domain.model.RacingResult;
+
+public record RacingGame(RaceCount raceCount, Cars cars) {
```
record로 구현 👍

javajigi commented on src/main/java/racinggame/domain/business/RacingGame.java:
```diff
@@ -0,0 +1,32 @@
+    }
+    
+    public boolean isEnd() {
+        return !raceCount.isZeroOrLower();
```
```suggestion
        return raceCount.isPositive();
```
부정문보다 메서드 추가해 구현하는 것을 추천

javajigi commented on src/main/java/racinggame/domain/model/Car.java:
```diff
@@ -0,0 +1,52 @@
+package racinggame.domain.model;
+
+public class Car {
```
👍

javajigi commented on src/main/java/racinggame/domain/model/Cars.java:
```diff
@@ -0,0 +1,58 @@
+
+public class Cars {
+    
+    List<Car> carList;
```
```suggestion
    private final List<Car> carList;
```

javajigi commented on src/main/java/racinggame/domain/model/Cars.java:
```diff
@@ -0,0 +1,58 @@
+                max = car.findLocation();
+            }
+        }
+        return max;
```
indent 2
객체 설계 충분히 연습했으니 stream api 사용해 구현해 본다.

javajigi commented on src/main/java/racinggame/domain/model/Cars.java:
```diff
@@ -0,0 +1,58 @@
+                carsName.add(car.showName());
+            }
+        }
+        return carsName;
```
indent 2
객체 설계 충분히 연습했으니 stream api 사용해 구현해 본다.

javajigi commented on src/main/java/racinggame/domain/model/Cars.java:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+    
+    public List<String> findWinners() {
```
반환 값도 일급 콜렉션 적용해 보는 것 어떨까?

javajigi reviewed (changes requested):
5단계 첫 번째 리뷰 요청은 정말 단순한 구조였는데요.
두 번째 리뷰 요청 진행할 때 객체 분리 원칙들 적용해 깔끔하게 구현 잘 했네요. 💯 
4단계까지 구현한 코드와 비교해 너무 달라졌네요.
힘듦이 있었겠지만 꾸준히 도전하는 모습 보기 좋네요.
그리 큰 피드백은 아니지만 피드백 몇 개 남겨 request changes로 피드백해요.
로또 미션 진행하면서 잠시 짬을 내 보세요.
그리 오랜 시간이 걸리지 않을 겁니다.

Commit:
refactor
1. 응답 타입도 일급 컬렉션 사용으로 변경
2. ! 을제거하기 위해 메서드 로직 변경
3. IO에서 값 처리 로직 제거하기 위해 로직 이관

Commit:
refactor :
Cars 객체 내 반복문을 stream으로 변경하고 불필요한 Car 내부 메서드 제거

javajigi reviewed (approved):
지금까지 자동차 경주 게임 미션 진행하느라 넘 수고했어요. 👍 
쉽지 않은 과정이었는데 끝까지 마무리 잘 했네요. 
로또 2단계 리뷰 요청했던데요.
이 미션 마무리하고 로또 미션에 집중해요.
